lib/improver/ensemble_copula_coupling/ensemble_copula_coupling.py

```diff
@@ -0,0 +1,533 @@
+
+from improver.ensemble_calibration.ensemble_calibration_utilities import (
+    concatenate_cubes, convert_cube_data_to_2d, rename_coordinate)
+from improver.ensemble_copula_coupling.ensemble_copula_coupling_constants \
+    import bounds_for_ecdf, units_of_bounds_for_ecdf
+from improver.ensemble_copula_coupling.ensemble_copula_coupling_utilities \
+    import create_cube_with_percentiles, create_percentiles
+
+
+class GeneratePercentilesFromProbabilities(object):
+    """
+    Class for generating percentiles from probabilities.
+    In combination with the Ensemble Reordering plugin, this is Ensemble
+    Copula Coupling.
+    """
+
+    def __init__(self):
+        """
+        Initialise the class.
+        """
+        pass
+
+    def _add_bounds_to_thresholds_and_probabilities(
+            self, threshold_points, probabilities_for_cdf, bounds_pairing):
+        """
+        Padding of the lower and upper bounds for a given phenomenon for the
+        threshold_points, and padding of probabilities of 0 and 1 to the
+        forecast probabilities.
+
+        Parameters
+        ----------
+        threshold_points : Numpy array
+            Array of threshold values used to calculate the probabilities.
+        probabilities_for_cdf : Numpy array
+            Array containing the probabilities used for constructing an
+            empirical cumulative distribution function i.e. probabilities
+            below threshold.
+
+        Returns
+        -------
+        threshold_points : Numpy array
+            Array of threshold values padded with the lower and upper bound.
+        probabilities_for_cdf : Numpy array
+            Array containing the probabilities padded with 0 and 1 at each end.
+        bounds_pairing : Tuple
+            Lower and upper bound to be used as the ends of the
+            empirical cumulative distribution function.
+
+        """
+        lower_bound, upper_bound = bounds_pairing
+        threshold_points = np.insert(threshold_points, 0, lower_bound)
+        threshold_points = np.append(threshold_points, upper_bound)
+        zeroes_array = np.zeros((probabilities_for_cdf.shape[0], 1))
+        ones_array = np.ones((probabilities_for_cdf.shape[0], 1))
+        probabilities_for_cdf = np.concatenate(
+            (zeroes_array, probabilities_for_cdf, ones_array), axis=1)
+        return threshold_points, probabilities_for_cdf
+
+    def _probabilities_to_percentiles(
+            self, forecast_probabilities, percentiles, bounds_pairing):
+        """
+        Conversion of probabilities to percentiles through the construction
+        of an empirical cumulative distribution function. This is effectively
+        constructed by linear interpolation from the probabilities associated
+        with each threshold to a set of percentiles.
+
+        Parameters
+        ----------
+        forecast_probabilities : Iris cube
+            Cube with a probability_above_threshold coordinate.
```

Comment:
Actually, maybe I'm wrong, maybe your convert_cube_data_to_2d function takes care of this